parlai/agents/seq2seq/seq2seq.py

```diff
@@ -462,185 +293,102 @@ def share(self):
-                    valid_cands.append((i, v, [curr_lcs[j] for j in valid_c_inds]))
-                    cs = torch.LongTensor(cs)
-                    if self.use_cuda:
-                        cs = cs.cuda()
-                    cands.append(cs)
-
-        return xs, ys, labels, valid_inds, cands, valid_cands, is_training
-
-    def init_cuda_buffer(self, batchsize):
-        if self.use_cuda and not hasattr(self, 'buffer_initialized'):
-            try:
-                print('preinitializing pytorch cuda buffer')
-                bsz = self.opt.get('batchsize', batchsize)
-                maxlen = self.truncate or 180
-                dummy = torch.ones(bsz, maxlen).long().cuda()
-                sc = self.model(dummy, dummy)[1]
-                loss = self.criterion(sc.view(-1, sc.size(-1)), dummy.view(-1))
-                loss.backward()
-                self.buffer_initialized = True
-            except RuntimeError as e:
-                if 'out of memory' in str(e):
-                    m = ('CUDA OOM: Lower batch size (-bs) from {} or lower max'
-                         ' sequence length (-tr) from {}'.format(bsz, maxlen))
-                    raise RuntimeError(m)
-                else:
-                    raise e
-
-    def batch_act(self, observations):
-        batchsize = len(observations)
-        self.init_cuda_buffer(batchsize)
-        # initialize a table of replies with this agent's id
-        batch_reply = [{'id': self.getID()} for _ in range(batchsize)]
-
-        # convert the observations into batches of inputs and targets
-        # valid_inds tells us the indices of all valid examples
-        # e.g. for input [{}, {'text': 'hello'}, {}, {}], valid_inds is [1]
-        # since the other three elements had no 'text' field
-        xs, ys, labels, valid_inds, cands, valid_cands, is_training = self.vectorize(observations)
-
-        if xs is None:
-            # no valid examples, just return empty responses
-            return batch_reply
-
-        # produce predictions, train on targets if availables
-        cand_inds = [i[0] for i in valid_cands] if valid_cands is not None else None
-        predictions, cand_preds = self.predict(xs, ys, cands, cand_inds, is_training)
-
-        if is_training:
-            report_freq = 0
-        else:
-            report_freq = self.report_freq
-        if predictions is not None:
-            PaddingUtils.map_predictions(
-                predictions, valid_inds, batch_reply, observations,
-                self.dict, self.END_IDX, report_freq=report_freq, labels=labels,
-                answers=self.answers, ys=ys.data if ys is not None else None)
-
+        except RuntimeError as e:
+            # catch out of memory exceptions during fwd/bck (skip batch)
+            if 'out of memory' in str(e):
+                print('| WARNING: ran out of memory, skipping batch. '
```

Comment:
I wonder if there's a cleaner way. Maybe doing it in TorchAgent around the train_step call? It's such a frustratingly large try statement.

It's fine though.